discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs

```diff
@@ -132,59 +133,59 @@ impl<const KEY_PHRASE_SIZE: usize> Tokenizer<KEY_PHRASE_SIZE> {
         };
 
         encoding
-            .is_valid(self.tokenizer.vocab_size(), KEY_PHRASE_SIZE)
+            .is_valid(self.tokenizer.get_vocab_size(true), KEY_PHRASE_SIZE)
             .then(|| (encoding, key_phrases))
+            .ok_or_else(|| "invalid encoding".into())
     }
 }
 
 /// Decodes the tokenized words.
 ///
 /// Joins starting tokens with their continuing tokens. Everything which is not separated by
 /// whitespace is considered as continuation as well, e.g. punctuation. All words are lowercased.
-fn decode_words(
-    sequence: impl AsRef<str>,
-    offsets: impl AsRef<[Offsets]>,
-    overflowing: Option<impl AsRef<[BertEncoding<i64>]>>,
-) -> Vec<String> {
-    let sequence = sequence.as_ref();
-    let offsets = offsets.as_ref();
+fn decode_words(sequence: &str, offsets: &[Offsets], overflowing: &[BertEncoding]) -> Vec<String> {
     let mut words = Vec::<String>::with_capacity(offsets.len());
     let (word_start, word_end) = offsets.iter().fold(
         (0, 0),
-        |(word_start, word_end), &Offsets(token_start, token_end)| {
+        |(word_start, word_end), &(token_start, token_end)| {
             if word_end < token_start {
                 words.push(sequence[word_start..word_end].to_lowercase());
                 (token_start, token_end)
             } else {
-                (word_start, token_end)
+                (
+                    word_start,
+                    // tokenizers uses (0, 0) instead of (n, n) as offsets for special tokens even
+                    // if they aren't at the beginning of the sequence but at the n-th position
+                    token_end.max(word_end),
```

Comment:
let's say you have a sequence `"some words"` which get tokenized to `"[CLS]", "some", "word", "##s", "[SEP]"` with the offsets `(0, 0), (0, 4), (5, 9), (9, 10), (0, 0)`. then the loop basically checks if there is a gap in the offsets which is taken as a word boundary (this is the if-branch). if there is no gap, then the tokens are concatenated until we reach the next gap or the end (this is the else-branch).

now, this assumes monotonically increasing offset values. the only case where this assumption is broken by the upstream tokenizer, is when a `(0, 0)` follows on a non-zero offset-end (`token_end`). hence we take the max of the current `word_end` state (which gets updated every loop iteration) and the current `token_end`, which concludes in two cases:
- either a correct offset will change nothing from the non-max-code,
- or a wrong zero-offset will cause the old `word_end` to be carried on instead of set to zero. the new `word_end` will still be correct, because only special tokens have wrong offsets and special tokens always have length zero in the original sequence.